src/kiota/Authentication/GitHub/DeviceCode/AcessTokenProvider.cs

```diff
@@ -0,0 +1,92 @@
+            var tokenResponse = await PollForTokenAsync(deviceCodeResponse, cancellationToken);
+            return tokenResponse?.AccessToken ?? string.Empty;
+        }
+        return string.Empty;
+	}
+	private async Task<AccessCodeResponse?> PollForTokenAsync(GitHubDeviceCodeResponse deviceCodeResponse, CancellationToken cancellationToken)
+	{
+		var timeOutTask = Task.Delay(TimeSpan.FromSeconds(deviceCodeResponse.ExpiresInSeconds), cancellationToken);
+		var pollTask = Task.Run(async () => {
+			while(!cancellationToken.IsCancellationRequested) {
+				var tokenResponse = await GetTokenAsync(deviceCodeResponse, cancellationToken);
+				if(tokenResponse != null)
+					return tokenResponse;
+				await Task.Delay(TimeSpan.FromSeconds(deviceCodeResponse.IntervalInSeconds), cancellationToken);
+			}
+			return null;
+		}, cancellationToken);
+		var completedTask = await Task.WhenAny(timeOutTask, pollTask);
+		if(completedTask == timeOutTask)
+			throw new TimeoutException("The device code has expired.");
+		return await pollTask;
+	}
+	private async Task<AccessCodeResponse?> GetTokenAsync(GitHubDeviceCodeResponse deviceCodeResponse, CancellationToken cancellationToken)
+	{
+        if(string.IsNullOrEmpty(deviceCodeResponse.DeviceCode))
+            return null;
+		using var tokenRequest = new HttpRequestMessage(HttpMethod.Post, $"{BaseLoginUrl}/oauth/access_token") {
+			Content = new FormUrlEncodedContent(new Dictionary<string, string> {
+				{ "client_id", ClientId },
+				{ "device_code", deviceCodeResponse.DeviceCode },
+				{ "grant_type", "urn:ietf:params:oauth:grant-type:device_code" }
+			})
+		};
+		tokenRequest.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+		using var tokenResponse = await HttpClient.SendAsync(tokenRequest, cancellationToken);
+		tokenResponse.EnsureSuccessStatusCode();
+		var tokenContent = await tokenResponse.Content.ReadAsStringAsync(cancellationToken);
+		var result = JsonSerializer.Deserialize<AccessCodeResponse>(tokenContent);
+		if ("authorization_pending".Equals(result?.Error, StringComparison.OrdinalIgnoreCase))
+			return null;
+		else if (!string.IsNullOrEmpty(result?.Error))
+			throw new InvalidOperationException($"Error while getting token: {result.Error} - {result.ErrorDescription}");
+		else
+			return result;
+	}
+
+	private async Task<GitHubDeviceCodeResponse?> GetDeviceCodeAsync(CancellationToken cancellationToken) {
+		using var codeRequest = new HttpRequestMessage(HttpMethod.Post, $"{BaseLoginUrl}/device/code") {
+			Content = new FormUrlEncodedContent(new Dictionary<string, string> {
+				{ "client_id", ClientId },
+				{ "scope", Scope }
+			})
+		};
+		codeRequest.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+		using var codeResponse = await HttpClient.SendAsync(codeRequest, cancellationToken);
+		codeResponse.EnsureSuccessStatusCode();
+		var codeResponseContent = await codeResponse.Content.ReadAsStringAsync(cancellationToken);
```

Comment:
Similar thoughts to using `ReadFromJsonAsync<>` here

Reply:
done.